interfaces/apparmor/apparmor.go

```diff
@@ -37,46 +37,69 @@ import (
 	"github.com/snapcore/snapd/osutil"
 )
 
-// LoadProfile loads an apparmor profile from the given file.
+// LoadProfile loads apparmor profiles from the given files.
 //
 // If no such profile was previously loaded then it is simply added to the kernel.
 // If there was a profile with the same name before, that profile is replaced.
-func LoadProfile(fname string) error {
-	return loadProfile(fname, dirs.AppArmorCacheDir)
+func LoadProfiles(fnames []string) error {
+	return loadProfiles(fnames, dirs.AppArmorCacheDir)
 }
 
 // UnloadProfile removes the named profile from the running kernel.
 //
 // The operation is done with: apparmor_parser --remove $name
 // The binary cache file is removed from /var/cache/apparmor
-func UnloadProfile(name string) error {
-	return unloadProfile(name, dirs.AppArmorCacheDir)
+func UnloadProfiles(names []string) error {
+	return unloadProfiles(names, dirs.AppArmorCacheDir)
 }
 
-func loadProfile(fname, cacheDir string) error {
+func loadProfiles(fnames []string, cacheDir string) error {
+	if len(fnames) == 0 {
+		return nil
+	}
+
 	// Use no-expr-simplify since expr-simplify is actually slower on armhf (LP: #1383858)
-	args := []string{"--replace", "--write-cache", "-O", "no-expr-simplify", fmt.Sprintf("--cache-loc=%s", cacheDir)}
+	args := []string{"--replace", "--write-cache", "-O", "no-expr-simplify",
+		fmt.Sprintf("--cache-loc=%s", cacheDir)}
+
 	if !osutil.GetenvBool("SNAPD_DEBUG") {
 		args = append(args, "--quiet")
 	}
-	args = append(args, fname)
+	args = append(args, fnames...)
 
 	output, err := exec.Command("apparmor_parser", args...).CombinedOutput()
 	if err != nil {
-		return fmt.Errorf("cannot load apparmor profile: %s\napparmor_parser output:\n%s", err, string(output))
+		if len(fnames) > 1 {
+			// Revert possibly applied profiles. Note that this will always return
+			// an error as at least one of the profiles could not be loaded. We do
+			// not log it as this is anyway a paliative solution.
+			args := []string{"--remove"}
```

Comment:
This is incorrect.

We cannot remove a profile if the loading failed. When loading fails we still have the previous profile in place (because we use --replace). Running without a profile vs running under an older version of the profile are really strongly different.

I would like to ensure that, in larger context, snapd behaves the same way with and without the aggregation in place. In the past when setting up security of any profile failed it would bubble up to eventually fail the task and run undo handlers. This must be preserved IMO.

As such, please don't remove loaded profiles here and just return the error.